src/Npgsql/NpgsqlConnector.Auth.cs

```diff
@@ -90,29 +93,89 @@ async Task AuthenticateSASL(List<string> mechanisms, bool async)
+            using (var sha256 = SHA256.Create())
+                storedKey = sha256.ComputeHash(clientKey);
+
+            var clientFirstMessageBare = "n=*,r=" + clientNonce;
+            var serverFirstMessage = $"r={firstServerMsg.Nonce},s={firstServerMsg.Salt},i={firstServerMsg.Iteration}";
+            var clientFinalMessageWithoutProof = "c=biws,r=" + firstServerMsg.Nonce;
+
+            var authMessage = $"{clientFirstMessageBare},{serverFirstMessage},{clientFinalMessageWithoutProof}";
+
+            var clientSignature = HMAC(storedKey, authMessage);
+            var clientProofBytes = XOR(clientKey, clientSignature);
+            var clientProof = Convert.ToBase64String(clientProofBytes);
+
+            var serverKey = HMAC(saltedPassword, ServerKey);
+            var serverSignature = HMAC(serverKey, authMessage);
+
+            var messageStr = $"{clientFinalMessageWithoutProof},p={clientProof}";
+
+            await WriteSASLResponse(Encoding.UTF8.GetBytes(messageStr), async);
+            await Flush(async);
 
             var saslFinalServerMsg = Expect<AuthenticationSASLFinalMessage>(await ReadMessage(async));
             if (saslFinalServerMsg.AuthRequestType != AuthenticationRequestType.AuthenticationSASLFinal)
                 throw new NpgsqlException("[SASL] AuthenticationSASLFinal message expected");
             var scramFinalServerMsg = new AuthenticationSCRAMServerFinalMessage(saslFinalServerMsg.Payload);
 
-            if (scramFinalServerMsg.ServerSignature != Convert.ToBase64String(scramFinalClientMsg.ServerSignature))
+            if (scramFinalServerMsg.ServerSignature != Convert.ToBase64String(serverSignature))
                 throw new NpgsqlException("[SCRAM] Unable to verify server signature");
 
             var okMsg = Expect<AuthenticationRequestMessage>(await ReadMessage(async));
             if (okMsg.AuthRequestType != AuthenticationRequestType.AuthenticationOk)
                 throw new NpgsqlException("[SASL] Expected AuthenticationOK message");
 
-            string GetNonce()
+            static string GetNonce()
             {
                 var nonceLength = 18;
-                var rncProvider = RandomNumberGenerator.Create();
-                var nonceBytes = new byte[nonceLength];
-                rncProvider.GetBytes(nonceBytes);
-                return Convert.ToBase64String(nonceBytes);
+                using (var rncProvider = RandomNumberGenerator.Create())
+                {
+                    var nonceBytes = new byte[nonceLength];
+                    rncProvider.GetBytes(nonceBytes);
+                    return Convert.ToBase64String(nonceBytes);
+                }
+            }
+
+            static byte[] Hi(string str, byte[] salt, int count)
+            {
+                using (var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(str)))
+                {
+                    var salt1 = new byte[salt.Length + 4];
+                    byte[] hi, u1;
```

Comment:
Here you can use them too.

Reply:
`HMACSHA256.ComputeHash()` doesn't seem to have an overload that accepts a Span (even on netstandard2.1). We could still do the base64 decoding into a larger array where we would set the last byte - instead of copying - but again, I'm not sure it's worth the effort in this code path...

Reply:
It has `TryComputeHash`, but in .NET Standard 2.1. Okay, let's leave it as is.